File: VisualGGPK/DatWrapper.cs
using LibDat;
using LibDat.Data;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Windows;

namespace VisualGGPK
{
    public class DatString
    {
        private readonly AbstractData _data;
        private readonly bool _isUser;

        public string DataType { get { return _data.GetType().ToString().Replace("LibDat.Data.", ""); } }

        public int OffsetStart { get { return _data.Offset; } }

        public int OffsetEnd { get { return _data.Offset + _data.Length; } }

        public int Length { get { return _data.Length; } }

        public bool IsUser { get { return _isUser; } }

        public string NewValue
        {
            get
            {
                var str = _data as StringData;
                return (str == null ? "" : str.NewValue);
            }
            set
            {
                var str = _data as StringData;
                if (str != null)
                    str.NewValue = value;
            }
        }

        public string Value { get { return _data.GetValueString(); } }

        public DatString(AbstractData data, bool isUser)
        {
            _data = data;
            _isUser = isUser;
        }
    }

    public class DatWrapper
    {
        private readonly string _datName;


        private DatContainer _dat;

        public RecordInfo RecordInfo
        {
            get { return _dat.RecordInfo; }
        }

        public long DataSectionffset
        {
            get { return DatContainer.DataSectionOffset; }
        }

        public long DataSectionDataLength
        {
            get { return _dat.DataSectionDataLength; }

        }
        public List<RecordData> Records
        {
            get { return _dat.Records; }
        }
        public Dictionary<int, AbstractData> DataEntries
        {
            get { return DatContainer.DataEntries; }
        }

        private readonly List<DatString> _dataStrings = new List<DatString>();
        public List<DatString> Strings
        {
            get { return _dataStrings; }
        }

        public DatWrapper(string fileName)
        {
            if (Path.GetExtension(fileName) == ".dat64") _datName = Path.GetFileName(fileName);
            else _datName = Path.GetFileNameWithoutExtension(fileName);
            var fileBytes = File.ReadAllBytes(fileName);

            using (var ms = new MemoryStream(fileBytes))
            {
                ParseDatFile(ms);
            }
        }

        public DatWrapper(Stream inStream, string fileName)
        {
            if (Path.GetExtension(fileName) == ".dat64") _datName = Path.GetFileName(fileName);
            else _datName = Path.GetFileNameWithoutExtension(fileName);
            ParseDatFile(inStream);
        }

        private void ParseDatFile(Stream inStream)
        {
            try
            {
                _dat = new DatContainer(inStream, _datName);

                var containerData = DataEntries.ToList();
                var userStringOffsets = _dat.GetUserStringOffsets();
                foreach (var keyValuePair in containerData)
                {
                    var data = keyValuePair.Value;
                    var isUser = userStringOffsets.Contains(keyValuePair.Key);
                    Strings.Add(new DatString(data, isUser));
                }
            }
            catch (Exception ex)
            {
                throw new Exception(string.Format(Settings.Strings["DatWrapper_ParseDatFile_Failed"], ex.ToString()), ex);
            }
        }

        public void Save(string savePath)
        {
            try
            {
                _dat.Save(savePath);
            }
            catch (Exception ex)
            {
                var errorString = new StringBuilder();

                var temp = ex;
                while (temp != null)
                {
                    errorString.AppendLine(temp.Message);
                    temp = temp.InnerException;
                }

                MessageBox.Show(string.Format(Settings.Strings["DatWrapper_Save_Failed"], errorString), Settings.Strings["Error_Caption"], MessageBoxButton.OK, MessageBoxImage.Error);
            }

            MessageBox.Show(string.Format(Settings.Strings["DatWrapper_Save_Successful"], savePath), Settings.Strings["DatWrapper_Save_Successful_Caption"], MessageBoxButton.OK, MessageBoxImage.Information);
        }

        public string GetCSV()
        {
            return _dat.GetCsvFormat();
        }
    }
}

File: DatEditer/DatWrapper.vb
Imports System.Collections.Generic
Imports System.IO
Imports System.Linq
Imports System.Text
Imports LibDat
Imports LibDat.Data

Namespace VisualGGPK
    Public Class DatString
        Private _data As AbstractData

        Public ReadOnly Property DataType As String
            Get
                Return _data.GetType().ToString().Replace("LibDat.Data.", "")
            End Get
        End Property

        Public ReadOnly Property OffsetStart As Integer
            Get
                Return _data.Offset
            End Get
        End Property

        Public ReadOnly Property OffsetEnd As Integer
            Get
                Return _data.Offset + _data.Length
            End Get
        End Property

        Public ReadOnly Property Length As Integer
            Get
                Return _data.Length
            End Get
        End Property

        Public ReadOnly Property IsUser As Boolean

        Public Property NewValue As String
            Get
                Dim str = TryCast(_data, StringData)
                If Not IsNothing(str) Then
                    Return str.NewValue
                Else
                    Return ""
                End If
            End Get
            Set(value As String)
                DirectCast(_data, StringData).NewValue = value
            End Set
        End Property

        Public ReadOnly Property Value As String
            Get
                Return _data.GetValueString()
            End Get
        End Property

        Public Sub New(Data As AbstractData, isUser As Boolean)
            _data = Data
            Me.IsUser = isUser
        End Sub
    End Class

    Public Class DatWrapper

        Private _datName As String
        Private _dat As DatContainer

        Public ReadOnly Property RecordInfo As RecordInfo
            Get
                Return _dat.RecordInfo
            End Get
        End Property

        Public ReadOnly Property DataSectionffset As Long
            Get
                Return DatContainer.DataSectionOffset
            End Get
        End Property

        Public ReadOnly Property DataSectionDataLength As Long
            Get
                Return _dat.DataSectionDataLength
            End Get
        End Property

        Public ReadOnly Property Records As List(Of RecordData)
            Get
                Return _dat.Records
            End Get
        End Property

        Public ReadOnly Property DataEntries As Dictionary(Of Integer, AbstractData)
            Get
                Return DatContainer.DataEntries
            End Get
        End Property

        Private _dataStrings As New List(Of DatString)
        Public ReadOnly Property Strings As List(Of DatString)
            Get
                Return _dataStrings
            End Get
        End Property

        Public Sub New(fileName As String)
            _datName = Path.GetFileNameWithoutExtension(fileName)
            Dim fileBytes = File.ReadAllBytes(fileName)
            ParseDatFile(New MemoryStream(fileBytes))
        End Sub

        Public Sub New(inStream As Stream, fileName As String)
            _datName = Path.GetFileNameWithoutExtension(fileName)
            ParseDatFile(inStream)
        End Sub

        Private Sub ParseDatFile(inStream As Stream)
            Try
                _dat = New DatContainer(inStream, _datName)
                Dim containerData = DataEntries.ToList()
                Dim userStringOffsets = _dat.GetUserStringOffsets()
                For Each keyValuePair In containerData
                    Dim data = keyValuePair.Value
                    Dim isUser = userStringOffsets.Contains(keyValuePair.Key)
                    Strings.Add(New DatString(data, isUser))
                Next
            Catch ex As Exception
                Throw New Exception(String.Format("Failed to read dat: {0}", ex.Message), ex)
            End Try
        End Sub

        Public Sub Save(savePath As String)
            Try
                _dat.Save(savePath)
            Catch ex As Exception
                Dim errorString As New StringBuilder()
                Dim temp = ex
                While (temp IsNot Nothing)
                    errorString.AppendLine(temp.Message)
                    temp = temp.InnerException
                End While
                MessageBox.Show(String.Format("Failed to save: {0}", errorString), "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            End Try
            MessageBox.Show(String.Format("Saved '{0}'", savePath), "Saved", MessageBoxButtons.OK, MessageBoxIcon.Information)
        End Sub

        Public Function GetCSV() As String
            Return _dat.GetCsvFormat()
        End Function

    End Class
End Namespace